packages/slate-commons/src/commands/replaceNode.tsx

```diff
@@ -1,27 +1,35 @@
+import { isElementNode } from '@prezly/slate-types';
+import type { Node, NodeMatch, Path } from 'slate';
 import { Editor, Transforms } from 'slate';
-import type { NodeMatch, Node, NodeEntry } from 'slate';
+
+import { replaceChildren } from './replaceChildren';
 
 interface Options<T extends Node> {
-    entry: NodeEntry<T>;
+    at?: Path;
     match: NodeMatch<T>;
+    select?: boolean;
 }
 
 export function replaceNode<Original extends Node, New extends Node>(
     editor: Editor,
-    options: Options<Original>,
     newNode: New,
+    options: Options<Original>,
 ) {
+    const { at, match, select = false } = options;
     Editor.withoutNormalizing(editor, () => {
-        const [node, path] = options.entry;
-
-        Transforms.unsetNodes<Original>(editor, Object.keys(node), {
-            at: path,
-            match: options.match,
-        });
-
-        Transforms.setNodes<Original | New>(editor, newNode, {
-            at: path,
-            match: options.match,
-        });
+        for (const [node, path] of Editor.nodes(editor, { at, match, mode: 'highest' })) {
+            Transforms.unsetNodes<Original>(editor, Object.keys(node), { at: path });
+            Transforms.setNodes<Original | New>(editor, newNode, { at: path });
+            if (isElementNode(newNode)) {
+                replaceChildren(editor, [node, path], newNode.children);
+            }
+            if (select) {
+                Transforms.select(editor, path);
+            }
+            // exit after the fist iteration, otherwise there's risk that we'll plant
+            // multiple instances of the same object into the document. Which is bad:
+            // Slate is relying on referential equality and uniqueness internally.
+            return;
```

Comment:
If we only need to iterate once, wouldn't it be cleaner to just get the first item from `Editor.nodes(...)`?

Reply:
Somehow it looks cleaner to me this way, as it naturally handles the zero results case without extra hassle  :thinking: 

```ts
for (const [node, path] of Editor.nodes(...)) {
    // ...
    return // ... ;
}
```

vs 

```ts
const [entry] = Editor.nodes(/* ... */);
if (entry) {
   const [node, path] = entry;
   // ...
}
```

Don't you think it's more elegant with the interrupted loop? Or am I absusing it? :)

Reply:
The zero results case makes sense, and you do save one line, bit I still thinks that's a bit hacky ;)

You can also return early `if (!entry)`, and then the code would be pretty elegant 🤔

Reply:
> You can also return early if (!entry), and then the code would be pretty elegant 🤔

That's debatable :) 

> [...] but I still thinks that's a bit hacky ;)

Ok, changing to an `if` then :+1:

Reply:
931f99d4192bc9dad2c124005bd34eac4d580060